src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,209 @@
+use crate::helpers::fabric::{ChannelId, MessageChunks, MessageEnvelope};
+use crate::helpers::Identity;
+use crate::net::server::MpcServerError;
+use crate::net::RecordHeaders;
+use crate::protocol::{QueryId, RecordId, UniqueStepId};
+use async_trait::async_trait;
+use axum::extract::{self, FromRequest, Query, RequestParts};
+use axum::http::Request;
+use axum::middleware::Next;
+use axum::response::Response;
+use hyper::Body;
+use tokio::sync::mpsc;
+
+/// Used in the axum handler to extract the `query_id` and `step` from the path of the request
+pub struct Path(QueryId, UniqueStepId);
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for Path {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, UniqueStepId)>::from_request(req).await?;
+        Ok(Path(query_id, step))
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+pub struct IdentityQuery {
+    identity: Identity,
+}
+
+/// After an [`OwnedPermit`] has been reserved, it can be used once to send an item on the channel.
+/// Panics if cloned while a permit exists.
+pub struct ReservedPermit<T>(Option<mpsc::OwnedPermit<T>>);
+
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn new(permit: mpsc::OwnedPermit<T>) -> Self {
+        Self(Some(permit))
+    }
+    /// # Panics
+    /// if called more than once
+    pub fn send(&mut self, item: T) {
+        self.0
+            .take()
+            .expect("should only call `send` once")
+            .send(item);
+    }
+}
+
+impl<T> Clone for ReservedPermit<T> {
+    /// # Panics
+    /// if a permit exists
+    fn clone(&self) -> Self {
+        assert!(self.0.is_none());
+        Self(None)
+    }
+}
+
+pub async fn message_stream_middleware<T: Send + 'static, B>(
+    sender: mpsc::Sender<T>,
+    mut req: Request<B>,
+    next: Next<B>,
+) -> Result<Response, MpcServerError> {
+    let permit = sender.reserve_owned().await?;
+    req.extensions_mut().insert(ReservedPermit::new(permit));
+    Ok(next.run(req).await)
+}
+
+/// extracts the [`MessageChunks`] from the request and forwards it to the Message layer via the
+/// `permit`. If we try to extract the `permit` via the `Extension`'s `FromRequest` implementation,
+/// it will call `.clone()` on it, which will remove the `OwnedPermit`. Thus, we must access the
+/// `permit` via `Request::extensions_mut`, which returns [`Extensions`] without cloning.
+/// Unfortunately, this also means we must wrap the `permit` in an `Option` such that we can
+/// `take()` the `permit`, freeing up the `req` for `Request::into_body`
```

Comment:
remove the part about cloning?

Reply:
It's probably important that we document the messiness with `FromRequest`. where do you recommend I have that? should it just be a regular comment within the fn?

Reply:
sorry I meant the part about `Option.take`

Reply:
gotcha. removed that part